### PR TITLE
CAS.o: Add some documentation for CASObjectFormats

### PR DESCRIPTION
**Pull request by dexonsmith**
Just a start; hopefully enough to be useful though. Expectation is that we keep this up-to-date as we make changes.

### CODE REVIEW

**Comment by dexonsmith** [comment]
I don't think this hooks up sphinx to actually build the docs; for now they're just markdown documents available in-tree.

@cachemeifyoucan , if you'd like to document FlatV1 differently / more detail / etc., I'm thinking you can follow up in tree. WDYT?

File: llvm/docs/CASObjectFormats/CASObjectFormats.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+There are major caveats that prevent the current formats from being useful in
+from being useful in practice. These mostly stem from them handling
+serialization to/from `jitlink::LinkGraph` and not much else. That said,
+they're useful for experimenting with encoding schemes and demonstrating
+storage optimization headroom.
+
+Not clear whether we'll evolve one of these or create a new one.
+
+### Caveat: Does not support real algorithms
```
Comment (cachemeifyoucan) [inline llvm/docs/CASObjectFormats/CASObjectFormats.md]: This feels more like just we don't have an interface that compiler or linker can call into.
LinkGraph totally supports linking concepts and FlatV1 can totally do a fair amount of lazy loading if needed. We just don't have API for those.

File: llvm/docs/CASObjectFormats/FlatV1.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+See also "Major caveats" section in CASObjectFormats.md.
+
+## Structure
+
+The top-level CAS object is called *compile-unit*. This top-level object is
+effectively the whole object file, with description templates for various
+concepts outlined to reduce aggregate storage cost.
+
+**compile-unit**: all graph information stored at the top-level
+- Data: all symbols, blocks, sections, referring to templates with a 4B index
```
Comment (cachemeifyoucan) [inline llvm/docs/CASObjectFormats/FlatV1.md:48]: Compile-unit in flatv1 is basically:
* Template array: CASIDs to other CAS objects that stores Symbol, Blocks, etc.
* Data: an array of indexes that encode all the references from one entities to another. The array works based on the fact that there is a deterministic traversal order to visit all the sections, symbols and blocks and each of them know how many indexes they need when visiting.

**Review by cachemeifyoucan — APPROVED** [review]
I think this is good and we can improve in tree.

File: llvm/docs/CASObjectFormats/CASObjectFormats.md (hunk 1 above)
Comment (dexonsmith) [reply llvm/docs/CASObjectFormats/CASObjectFormats.md]: Rewored to "Not ready to support" and clarified text a bit.

File: llvm/docs/CASObjectFormats/FlatV1.md (hunk 2 above)
Comment (dexonsmith) [reply llvm/docs/CASObjectFormats/FlatV1.md:48]: Thanks; cleaned up the description a bit!

**Pull request merged**